### DISCUSSION TITLE
请教大家如何在测试环境和生产之前进行工作流发布上线活动

### DISCUSSION THREAD

**Discussion by pppeace** [apache/dolphinscheduler#16753] [Q&A]
如题，目前我们的做法是：
1. 创建不同的项目用于区分环境，如：dev，pre，prod。也可以是不同的Dolphin（对应测试和生产集群）；
2. 在 dev 项目中工作，创建工作流及任务，并完成测试工作；
3. 将测试通过的工作流导出 workflow.json，然后导出到 pre 项目中。

遇到的问题有：
1. 跨集群场景，工作流中使用的**资源文件**（hdfs），无法被传递，而且通过 hdfs 手动上传到指定路径，Dolphin **不能直接识别**；
2. **SQL 节点**中，指定的**数据源实例**，由于切换了环境，也需要随之切换，目前只能手动一个一个修改；
3. 上下游的**子工作流节点**会变成一个 **id**（如下图），需要重新选择指明上下游任务；
![image](https://github.com/user-attachments/assets/9a7f9485-2df3-45d6-8fce-59156d68c380)
4. 跨集群场景，work 分组/环境名称，需要手动创建对齐；

如上，因为手动环节较多，所以经常发生一些 miss，故希望像大家请教，看看应该如何更好的进行发布上线工作。

**Comment by reele** [comment]
自带的导入导出过于依赖code，对于跨环境的配置同步不友好，尤其是有依赖节点的情况下

我的解决方案是弃用自带的导入导出，依赖外部配置，把配置通过版本管理同步到生产环境，用脚本调用API实现外部配置的导入

这是我用于导入调度配置的脚本，你可以参考下：
https://github.com/reele/process_definition_import_tool
